core/src/commands/on_runtime_upgrade.rs

```diff
@@ -120,43 +125,93 @@ where
             log::info!(
                 "🔬 TryRuntime_on_runtime_upgrade succeeded! Running it again without checks for weight measurements."
             );
-            let (_, proof, encoded_result) = state_machine_call_with_proof::<Block, HostFns>(
+            let (proof, encoded_result) = state_machine_call_with_proof::<Block, HostFns>(
                 &ext,
+                &mut Default::default(),
                 &executor,
                 "TryRuntime_on_runtime_upgrade",
                 UpgradeCheckSelect::None.encode().as_ref(),
                 Default::default(), // we don't really need any extensions here.
-                shared.export_proof,
+                shared.export_proof.clone(),
             )?;
             let ref_time_results = encoded_result.try_into()?;
             (proof, ref_time_results)
         }
     };
 
+    // Check idempotency
+    let idempotency_ok = match command.no_idempotency_checks {
+        true => {
+            log::info!("ℹ Skipping idempotency check");
+            true
+        }
+        false => {
+            log::info!(
+                "🔬 Running TryRuntime_on_runtime_upgrade again to check idempotency: {:?}",
+                command.checks
+            );
+            let (oc_pre_root, _) = overlayed_changes.storage_root(&ext.backend, ext.state_version);
+            match state_machine_call_with_proof::<Block, HostFns>(
+                &ext,
+                &mut overlayed_changes,
+                &executor,
+                "TryRuntime_on_runtime_upgrade",
+                command.checks.encode().as_ref(),
```

Comment:
let's use literally same command arguments as above, i.e. `UpgradeCheckSelect::None.encode().as_ref()`

Reply:
why? if the dev is running with pre/post checks, we want to check that `pre_upgrade` and `post_upgrade` don't fail the second time they are executed

Reply:
oh, sorry, had some blackout, I should have asked to use `command.checks.encode().as_ref()` above as well; you want to run pre/post checks during the actual migration

Reply:
no worries :)